src/Lean/Elab/Tactic/BVDecide/LRAT/Internal/Clause.lean

```diff
@@ -0,0 +1,418 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Josh Clune
+-/
+prelude
+import Init.Data.List.Erase
+import Std.Sat.CNF.Basic
+import Lean.Elab.Tactic.BVDecide.LRAT.Internal.PosFin
+import Lean.Elab.Tactic.BVDecide.LRAT.Internal.Assignment
+
+
+namespace Lean.Elab.Tactic.BVDecide
+namespace LRAT
+namespace Internal
+
+open Std.Sat
+
+/--
+An inductive datatype used specifically for the output of the `reduce` function. The intended
+meaning of each constructor is explained in the docstring of the `reduce` function.
+-/
+inductive ReduceResult (α : Type u)
+  | encounteredBoth
+  | reducedToEmpty
+  | reducedToUnit (l : Literal α)
+  | reducedToNonunit
+
+/--
+Typeclass for clauses. An instance `[Clause α β]` indicates that `β` is the type of a clause with
+variables of type `α`.
+-/
+class Clause (α : outParam (Type u)) (β : Type v) where
+  toList : β → CNF.Clause α
+  not_tautology : ∀ c : β, ∀ l : Literal α, l ∉ toList c ∨ Literal.negate l ∉ toList c
+  /-- Returns none if the given array contains complementary literals -/
+  ofArray : Array (Literal α) → Option β
+  ofArray_eq :
+    ∀ arr : Array (Literal α), (∀ i : Fin arr.size, ∀ j : Fin arr.size, i.1 ≠ j.1 → arr[i] ≠ arr[j]) →
+      ∀ c : β, ofArray arr = some c → toList c = arr.toList
+  empty : β
+  empty_eq : toList empty = []
+  unit : Literal α → β
+  unit_eq : ∀ l : Literal α, toList (unit l) = [l]
+  isUnit : β → Option (Literal α)
+  isUnit_iff : ∀ c : β, ∀ l : Literal α, isUnit c = some l ↔ toList c = [l]
+  negate : β → CNF.Clause α
+  negate_eq : ∀ c : β, negate c = (toList c).map Literal.negate
+  /-- Returns none if the result is a tautology. -/
+  insert : β → Literal α → Option β
+  delete : β → Literal α → β
+  delete_iff : ∀ c : β, ∀ l : Literal α, ∀ l' : Literal α,
+    l' ∈ toList (delete c l) ↔ l' ≠ l ∧ l' ∈ toList c
+  contains : β → Literal α → Bool
+  contains_iff : ∀ c : β, ∀ l : Literal α, contains c l ↔ l ∈ toList c
+  /-- Reduces the clause with respect to the given assignment -/
+  reduce : β → Array Assignment → ReduceResult α
+
+namespace Clause
+
+instance : Entails α (Literal α) where
+  eval := fun p l => p l.1 = l.2
+
+instance (p : α → Bool) (l : Literal α) : Decidable (p ⊨ l) :=
+  inferInstanceAs (Decidable (p l.1 = l.2))
+
+def eval [Clause α β] (a : α → Bool) (c : β) : Bool :=
+  (toList c).any fun (l : Literal α) => a ⊨ l
+
+instance [Clause α β] : Entails α β where
+  eval a c := Clause.eval a c
+
+instance [Clause α β] (p : α → Bool) (c : β) : Decidable (p ⊨ c) :=
+  inferInstanceAs (Decidable (Clause.eval p c = true))
+
+instance [Clause α β] : Inhabited β where
+  default := empty
+
+end Clause
+
+/--
+The `DefaultClause` structure is primarily a list of literals. The additional field `nodupkey` is
+included to ensure that `not_tautology` is provable (which is needed to prove `insertRup_entails_hsat`
+and `insertRat_entails_hsat` in `LRAT.Formula.RupAddSound.lean` and `LRAT.Formula.RatAddSound.lean`).
```

Comment:
I think you want either `LRAT/Formula/Internal/RupAddSound.lean` or `LRAT.Formula.Internal.RupAddSound` here?